src/settings/nm-settings.c

```diff
@@ -896,16 +880,27 @@ connection_removed (NMSettingsConnection *connection, gpointer user_data)
 		g_signal_handlers_disconnect_by_func (connection, G_CALLBACK (connection_ready_changed), self);
 	g_object_unref (self);
 
-	/* Forget about the connection internally */
-	g_hash_table_remove (priv->connections, (gpointer) cpath);
 	_clear_connections_cached_list (&priv->connections_cached_list);
 
-	_emit_connection_removed (self, connection);
+	/* we unref @connection below. */
+	g_hash_table_steal (priv->connections, (gpointer) cpath);
 
-	/* Re-emit for listeners like NMPolicy */
-	_notify (self, PROP_CONNECTIONS);
-	if (nm_dbus_object_is_exported (NM_DBUS_OBJECT (connection)))
-		nm_dbus_object_unexport (NM_DBUS_OBJECT (connection));
+	if (priv->connections_loaded) {
+		_notify (self, PROP_CONNECTIONS);
+
+		nm_dbus_object_emit_signal (NM_DBUS_OBJECT (self),
+		                            &interface_info_settings,
+		                            &signal_info_connection_removed,
+		                            "(o)",
+		                            nm_dbus_object_get_path (NM_DBUS_OBJECT (connection)));
+	}
+
+	nm_dbus_object_unexport (NM_DBUS_OBJECT (connection));
+
+	if (priv->connections_loaded)
+		g_signal_emit (self, signals[CONNECTION_REMOVED], 0, connection);
+
+	g_object_ref (connection);
```

Comment:
Can you add a comment explaining why you change the order? It's not clear to me.

Reply:
Hm... the order is anyway something we define, so that it makes the most sense. For example, it makes (more) sense (to me) emitting `PropertyChanged` signal on `NMSettings:Connections`, before emitting `NMSettings:ConnectionRemoved` signal on D-Bus. I changed it, because I think the order as it is now, makes the most sense considering all the alternatives.

(note that before the D-Bus rework, the `NMSettings:Connections` property changed signal would always be sent on an idle handler, so it wasn't even (realistically) possible to emitting the D-Bus signal before `NMSettings:ConnectionRemoved` signal. Now it is, and that was changed).

Let's not look at this intermediate commit (because it was changed afterwards again, sorry). But for example, previously we would emit the internal signal `NM_SETTINGS_SIGNAL_CONNECTION_REMOVED` while the connection is still exported on D-Bus. I think, the other order is better, because AFAIS, the signal handlers care about doing additional cleanup after the connection is mostly already gone.

I don't have exact reasons to comment why the order should be exactly like that, except, it feels more right to me. So, don't know what to comment beyond explaining what the order of events *is*. That is, I cannot easily explain *why* the order is that way, only a trival comment saying that it *is* that way. So, I'd rather not add a comment.

Does that answer the question? Do you disagree with the *is* and have a stronger *why* then I can provide?

Reply:
Okay